## LAB 2.2 A1: Testing Pretrained-CNNs

### Proceedure:

- Using Dogs-Cats-Pandas
- Test and time **Xception**
- Test and time **MobileNet**
- Test and time **EfficientNetB0**
- Compare all with previous custom CNN

In [11]:
import tensorflow as tf, time, numpy as np
from tensorflow.keras.preprocessing import image_dataset_from_directory

train_ds = image_dataset_from_directory(
    "data/dogs_cats_pandas/animals",
    validation_split=0.2,
    subset="training",
    seed=123,
    image_size=(224, 224),
    batch_size=32
)

val_ds = image_dataset_from_directory(
    "data/dogs_cats_pandas/animals",
    validation_split=0.2,
    subset="validation",
    seed=123,
    image_size=(224, 224),
    batch_size=32
)

train_ds = train_ds.prefetch(buffer_size=tf.data.AUTOTUNE)
val_ds = val_ds.prefetch(buffer_size=tf.data.AUTOTUNE)
num_classes = 3

sample_batch = next(iter(val_ds))[0][:32]

Found 3000 files belonging to 3 classes.
Using 2400 files for training.
Found 3000 files belonging to 3 classes.
Using 600 files for validation.


In [12]:
t0 = time.time()
base = tf.keras.applications.Xception(weights="imagenet", include_top=False, input_shape=(224,224,3))
base.trainable = False
model = tf.keras.Sequential([
    base,
    tf.keras.layers.GlobalAveragePooling2D(),
    tf.keras.layers.Dense(num_classes, activation="softmax")
])
model.compile(optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"])
load_time = time.time()-t0
history = model.fit(train_ds, validation_data=val_ds, epochs=3, verbose=0)
t0 = time.time()
_ = model.predict(sample_batch, verbose=0)
pred_time = (time.time()-t0)/len(sample_batch)
print("Xception Accuracy:", history.history["val_accuracy"][-1])
print("Load time:", load_time)
print("Avg predict time:", pred_time)

Xception Accuracy: 0.5883333086967468
Load time: 1.1828858852386475
Avg predict time: 0.0162665918469429


In [13]:
t0 = time.time()
base = tf.keras.applications.MobileNet(weights="imagenet", include_top=False, input_shape=(224,224,3))
base.trainable = False
model = tf.keras.Sequential([
    base,
    tf.keras.layers.GlobalAveragePooling2D(),
    tf.keras.layers.Dense(num_classes, activation="softmax")
])
model.compile(optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"])
load_time = time.time()-t0
history = model.fit(train_ds, validation_data=val_ds, epochs=3, verbose=0)
t0 = time.time()
_ = model.predict(sample_batch, verbose=0)
pred_time = (time.time()-t0)/len(sample_batch)
print("MobileNet Accuracy:", history.history["val_accuracy"][-1])
print("Load time:", load_time)
print("Avg predict time:", pred_time)

MobileNet Accuracy: 0.7366666793823242
Load time: 0.4708428382873535
Avg predict time: 0.008874975144863129


In [14]:
# cell 4 EfficientNetB0
t0 = time.time()
base = tf.keras.applications.EfficientNetB0(weights="imagenet", include_top=False, input_shape=(224,224,3))
base.trainable = False
model = tf.keras.Sequential([
    base,
    tf.keras.layers.GlobalAveragePooling2D(),
    tf.keras.layers.Dense(num_classes, activation="softmax")
])
model.compile(optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"])
load_time = time.time()-t0
history = model.fit(train_ds, validation_data=val_ds, epochs=3, verbose=0)
t0 = time.time()
_ = model.predict(sample_batch, verbose=0)
pred_time = (time.time()-t0)/len(sample_batch)
print("EfficientNetB0 Accuracy:", history.history["val_accuracy"][-1])
print("Load time:", load_time)
print("Avg predict time:", pred_time)


EfficientNetB0 Accuracy: 0.996666669845581
Load time: 1.5929367542266846
Avg predict time: 0.02901402860879898


### Comparison of CNNs on Dogs-Cats-Pandas

| | **Xception** | **MobileNet** | **EfficientNetB0** | **Custom CNN from LAB2_1_B** |
|---|---|---|---|---|
| Accuracy | 0.60 | 0.73 | 0.99 | 0.73 |
| Load time (s) | 1.18 | 0.47 | 1.59 | - |
| Avg. Time to Predict (s) | 0.016 | 0.008 | 0.029 | - |